zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,494 @@
+                    }
+                }
+            }
+        }
+    }
+
+    fn parse_host_port(port_str: &str) -> Option<u16> {
+        let parts: Vec<&str> = port_str.split(':').collect();
+        if parts.len() > 1 {
+            if let Some(host_port_str) = parts.get(parts.len() - 2) {
+                if let Ok(port) = host_port_str.parse::<u16>() {
+                    return Some(port);
+                }
+            }
+        }
+        None
+    }
+
+    fn add_parsed_port(
+        port: u16,
+        path: &str,
+        file_path: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) {
+        let description = format!("[{}] {}", file_path.display(), path);
+        ecosystem_ports.add_port_info(port, description);
+    }
+}
+
+pub trait ConfigWithChainPorts {
+    fn get_default_ports() -> HashMap<String, u16>;
+    fn set_ports(&mut self, ports: HashMap<String, u16>) -> Result<()>;
+}
+
+impl ConfigWithChainPorts for ExplorerBackendPorts {
+    fn get_default_ports() -> HashMap<String, u16> {
+        HashMap::from([
+            ("api_http_port".to_string(), DEFAULT_EXPLORER_API_PORT),
+            (
+                "data_fetcher_http_port".to_string(),
+                DEFAULT_EXPLORER_DATA_FETCHER_PORT,
+            ),
+            ("worker_http_port".to_string(), DEFAULT_EXPLORER_WORKER_PORT),
+        ])
+    }
+
+    fn set_ports(&mut self, ports: HashMap<String, u16>) -> anyhow::Result<()> {
+        if ports.len() != Self::get_default_ports().len() {
+            bail!("Incorrect number of ports provided");
+        }
+        for (desc, port) in ports {
+            match desc.as_str() {
+                "api_http_port" => self.api_http_port = port,
+                "data_fetcher_http_port" => self.data_fetcher_http_port = port,
+                "worker_http_port" => self.worker_http_port = port,
+                _ => bail!("Unknown port descriptor: {}", desc),
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ConfigWithChainPorts for GeneralConfig {
+    fn get_default_ports() -> HashMap<String, u16> {
+        HashMap::from([
+            (
+                "web3_json_rpc_http_port".to_string(),
+                DEFAULT_WEB3_JSON_RPC_PORT,
+            ),
+            (
+                "web3_json_rpc_ws_port".to_string(),
+                DEFAULT_WEB3_WS_RPC_PORT,
+            ),
+            ("healthcheck_port".to_string(), DEFAULT_HEALTHCHECK_PORT),
+            ("merkle_tree_port".to_string(), DEFAULT_MERKLE_TREE_PORT),
+            (
+                "prometheus_listener_port".to_string(),
+                DEFAULT_PROMETHEUS_PORT,
+            ),
+            (
+                "contract_verifier_port".to_string(),
+                DEFAULT_CONTRACT_VERIFIER_PORT,
+            ),
+            ("consensus_port".to_string(), DEFAULT_CONSENSUS_PORT),
+        ])
+    }
+
+    fn set_ports(&mut self, ports: HashMap<String, u16>) -> anyhow::Result<()> {
```

Comment:
Do I understand it right, that if we have some another port, which is not specified, it'd fail?

Reply:
In this case, i'd rather save the whole paths to the value and write a small macros for updating it

Reply:
`set_ports` is used [here](https://github.com/matter-labs/zksync-era/pull/2849/files#diff-7847b0da60e1ab89e4bb36e78a39f1e249133f2c82b78d86aecf1bbfd0b19d1eR60)

This means `set_ports` uses the same ports as the ones in `get_default_ports`
So it wouldn't fail as long as `get_default_ports` is consistent with `set_ports`. Both functions belong to the same trait: `ConfigWithChainPorts` making it easy to keep it consistent.

How do you think we can improve this?